domain/credential/state/state.go

```diff
@@ -507,3 +509,74 @@ func (st *State) WatchCredential(
 	result, err := getWatcher("cloud_credential", uuid, changestream.All)
 	return result, errors.Annotatef(err, "watching credential")
 }
+
+// GetCloud returns the cloud with the specified name.
+func (st *State) GetCloud(ctx context.Context, name string) (cloud.Cloud, error) {
+	db, err := st.DB()
+	if err != nil {
+		return cloud.Cloud{}, errors.Trace(err)
+	}
+
+	var result []cloud.Cloud
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		var err error
+		result, err = dbcloud.LoadClouds(ctx, tx, name)
+		return errors.Trace(err)
+	})
+	if len(result) == 0 {
+		return cloud.Cloud{}, fmt.Errorf("cloud %q %w", name, errors.NotFound)
+	}
+	return result[0], errors.Trace(err)
+}
+
+// ModelsUsingCloudCredential returns a map of uuid->name for models which use the credential.
+func (st *State) ModelsUsingCloudCredential(ctx context.Context, id credential.ID) (map[model.UUID]string, error) {
+	if err := id.Validate(); err != nil {
+		return nil, errors.Annotate(err, "invalid credential querying models")
+	}
+
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	query := `
+SELECT mm.model_uuid AS &M.model_uuid, mm.name AS &M.name
+FROM   model_metadata mm
+JOIN cloud_credential cc ON cc.uuid = mm.cloud_credential_uuid
+JOIN cloud ON cloud.uuid = cc.cloud_uuid
```

Comment:
This join doesn't get us anything.

Reply:
It's needed because of the WHERE condition which selects on cloud name.

Reply:
Apologies I missed this.